extensions/amp-apester-media/0.1/amp-apester-media.js

```diff
@@ -316,60 +318,59 @@ class AmpApesterMedia extends AMP.BaseElement {
-                const media = this.embedOptions_.playlist
-                  ? payload[Math.floor(Math.random() * payload.length)]
-                  : payload;
-                const src = this.constructUrlFromMedia_(media['interactionId']);
-                const iframe = this.constructIframe_(src);
-                const overflow = this.constructOverflow_();
-                const mutate = state => {
-                  state.element.classList.add('i-amphtml-apester-iframe-ready');
-                };
-                const state = {
-                  element: iframe,
-                  mutator: mutate,
-                };
-                this.mediaId_ = media['interactionId'];
-                this.iframe_ = iframe;
-                this.element.appendChild(overflow);
-                this.element.appendChild(iframe);
-                this.registerToApesterEvents_();
+    const vsync = Services.vsyncFor(this.win);
+    return this.queryMedia_().then(
+        response => {
+          const payload = response['payload'];
+          // If it's a playlist we choose a media randomly.
+          // The response will be an array.
+          const media = this.embedOptions_.playlist
+            ? payload[Math.floor(Math.random() * payload.length)]
+            : payload;
+          const src = this.constructUrlFromMedia_(media['interactionId']);
+          const iframe = this.constructIframe_(src);
+          this.intersectionObserverApi_ = new IntersectionObserverApi(
+              this,
+              iframe
+          );
+
+          this.mediaId_ = media['interactionId'];
+          this.iframe_ = iframe;
+          this.registerToApesterEvents_();
 
-                return (this.iframePromise_ = this.loadPromise(iframe)
-                    .then(() => {
-                      Services.vsyncFor(this.win).runPromise({mutate}, state);
-                      return media;
-                    }));
-              },
-              error => {
-                dev().error(TAG, 'Display', error);
-                return undefined;
-              }
-          )
-    /** @param {!JsonObject} media */
-          .then(media => {
-            this.togglePlaceholder(false);
-            this.ready_ = true;
-            let height = 0;
-            if (media && media['data'] && media['data']['size']) {
-              height = media['data']['size']['height'];
-            }
-            if (height != this.height_) {
-              this.height_ = height;
-              if (this.random_) {
-                this./*OK*/ attemptChangeHeight(height);
-              } else {
-                this./*OK*/ changeHeight(height);
-              }
-            }
-          })
+          return vsync.mutatePromise(() => {
+            const overflow = this.constructOverflow_();
+            this.element.appendChild(overflow);
+            this.element.appendChild(iframe);
+            return this.loadPromise(iframe).then(() => {
```

Comment:
`mutatePromise` does not resolve to any value (it's not a real promise), you'll have to chain a `then` callback.

```js
return vsync.mutatePromise(() => {
  // mutations go in here
}).then(() => {
  return this.loadPromise(iframe);
});

Reply:
:open_mouth: TIL. Can we make it resolve to the return of the `mutator`?

Reply:
Nope, we did a performance optimization a long time ago to reuse the same promise for calls to `mutatePromise`, which prevents return values. General case doesn't have a return value, so it's ok to just chain a `then`.

Reply:
Tthanks for explaining, some regex to capture `return` inside body of `mutatePromise` in presubmit seems doable.

Reply:
Hi guys, Thank you for the comments.
I need the Iframe to finish loading before I perform some of the actions.
Is it ok to add another loadPromise to the end of the chain?

```javascript
return this.loadPromise(iframe).then(() => { 
  return vsync.mutatePromise(() => {
    // mutations go in here
  }).then(() => {
    return this.loadPromise(iframe); 
  });
});
```

Reply:
Yup, that's fine.